tiledb/sm/metadata/test/unit_metadata.cc

```diff
@@ -0,0 +1,140 @@
+/**
+ * @file tiledb/sm/metadata/test/unit_metadata.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2021 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines a test `main()`
+ */
+
+#include <catch.hpp>
+#include "../metadata.h"
+#include "tiledb/common/dynamic_memory/dynamic_memory.h"
+#include "tiledb/sm/buffer/buffer.h"
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/filesystem/uri.h"
+#include "tiledb/sm/misc/time.h"
+#include "tiledb/sm/misc/uuid.h"
+
+using namespace tiledb;
+using namespace tiledb::common;
+using namespace tiledb::sm;
+
+template <class T, int n>
+inline T& buffer_metadata(void* p) {
+  return *static_cast<T*>(static_cast<void*>(static_cast<char*>(p) + n));
+}
+
+TEST_CASE(
+    "Metadata: Test metadata deserialization", "[metadata][deserialization]") {
+  std::vector<tdb_shared_ptr<Buffer>> metadata_buffs;
+
+  // key1:a, value1:100,200
+  std::string key1 = "key1";
+  uint32_t key1_size = static_cast<uint32_t>(key1.size());
+  uint32_t value1_size = 2;
```

Comment:
You've got the number of values listed here but not an array of values you're using. Better to initialize and check by index below.